### PR TITLE
Introduce Flow types.

### PR DESCRIPTION
**Pull request by ming-soon**
Refs #339

### CODE REVIEW

File: src/getDataTransferItems.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
-export default function getDataTransferFiles(event, isMultipleAllowed = true) {
+/* @flow */
+export default function getDataTransferFiles(event: Object, isMultipleAllowed: boolean = true) {
```
Comment (ryyppy) [inline src/getDataTransferItems.js]: What exactly is an `event` ? apparently you want to express an object that has a `event.dataTransfer` attribute and a `event.target` attribute... why not make this explicit?

File: src/index.js [hunk 2]
```diff
@@ -1,23 +1,84 @@
...
+  type: string,
+  style: Object,
+  multiple: boolean,
+  name?: string,
+  ref: Function,
+  onChange: Function
+};
+
+export type Props = {
+  onClick: Function,
```
Comment (ryyppy) [inline src/index.js]: what kind of Function is this? Make this explicit!
E.g. `onClick: (elementId: string) => void`

File: src/index.js [hunk 3]
```diff
@@ -1,23 +1,84 @@
...
     super(props, context);
-    this.onClick = this.onClick.bind(this);
-    this.onDragStart = this.onDragStart.bind(this);
-    this.onDragEnter = this.onDragEnter.bind(this);
-    this.onDragLeave = this.onDragLeave.bind(this);
-    this.onDragOver = this.onDragOver.bind(this);
-    this.onDrop = this.onDrop.bind(this);
-    this.onFileDialogCancel = this.onFileDialogCancel.bind(this);
-    this.fileAccepted = this.fileAccepted.bind(this);
+    (this:any).onClick = this.onClick.bind(this);
```
Comment (ryyppy) [inline src/index.js]: you need to define `onClick` as a property type in the class body, to make this flow understand the `this.onClick`...

File: src/index.js [hunk 4]
```diff
@@ -27,21 +88,25 @@ class Dropzone extends React.Component {
...
   componentWillUnmount() {
     // Can be replaced with removeEventListener, if addEventListener works
-    document.body.onfocus = null;
+    if (document.body instanceof HTMLElement) {
+      document.body.onfocus = null;
+    }
   }
 
-  onDragStart(e) {
+  onDragStart(e: Object) {
```
Comment (ryyppy) [inline src/index.js]: I guess `e` is a DOM element... you can leverage the internal flowtype libdefs, which define exactly this: `Element` as defined here: https://github.com/facebook/flow/blob/master/lib/dom.js#L939

Comment (okonet) [inline src/index.js]: `e` is `event`. Let's rename arguments to make them less ambiguous.

**Comment by okonet** [comment]
@ming-soon thanks for the PR! This is looking promising to me. Let's address feedback from @ryyppy and we're almost there I think.

One more concern is deprecating of some props. We should probably wait with this PR until next major release since I'd like to use https://github.com/Aweary/react-is-deprecated and I don't know how this will work with Flow. Ideas?

File: src/index.js (hunk 4 above)
Comment (ryyppy) [inline src/index.js]: okay, so it will be `onDragStart(e: Event)` then :-)

Comment (ming-soon) [inline src/index.js]: We can use `SyntheticDragEvent` to be more specific.

[https://github.com/facebook/flow/blob/v0.35.0/lib/react.js#L349](https://github.com/facebook/flow/blob/v0.35.0/lib/react.js#L349)

Comment (okonet) [inline src/index.js]: Yes, let's reuse React's type defs as much as possible.

File: .babelrc [hunk 5]
```diff
@@ -1,4 +1,4 @@
 {
   "presets": ["es2015", "react", "stage-0"],
-  "plugins": ["add-module-exports"]
+  "plugins": ["add-module-exports", "transform-flow-strip-types"]
```
Comment (okonet) [inline .babelrc]: I'm wondering how we should deliver flow types to npm so people can get advantage of them in their projects? @ryyppy I know you did this before.

File: src/getDataTransferItems.js (hunk 1 above)
Comment (okonet) [inline src/getDataTransferItems.js]: Let's add return types everythere for such cases: `function(arg: type): type {}`

File: src/index.js [hunk 6]
```diff
@@ -1,23 +1,84 @@
...
+
 const supportMultiple = (typeof document !== 'undefined' && document && document.createElement) ?
   'multiple' in document.createElement('input') :
   true;
 
 class Dropzone extends React.Component {
-  constructor(props, context) {
+  props: Props;
+
+  static defaultProps: {
```
Comment (okonet) [inline src/index.js]: Can we somehow co-locate this with https://github.com/okonet/react-dropzone/pull/344/files#diff-1fdf421c05c1140f6d71444ea2b27638R367?

File: src/index.js [hunk 7]
```diff
@@ -1,23 +1,84 @@
...
+    maxSize: number,
+    minSize: number
+  };
+
+  state: {
+    isDragActive: boolean,
+    isDragReject?: boolean
+  };
+
+  fileInputEl: Object;
```
Comment (okonet) [inline src/index.js]: This should be a React node, isn't it?

File: src/index.js (hunk 4 above)
Comment (okonet) [inline src/index.js]: Yes, please!

File: src/index.js [hunk 8]
```diff
@@ -27,21 +88,25 @@ class Dropzone extends React.Component {
...
 
-  onDragStart(e) {
+  onDragStart(e: Object) {
     if (this.props.onDragStart) {
       this.props.onDragStart.call(this, e);
     }
   }
 
-  onDragEnter(e) {
+  onDragEnter(e: Object) {
```
Comment (okonet) [inline src/index.js]: `onDragEnter(e: Object): void`?

File: src/index.js [hunk 9]
```diff
@@ -166,15 +231,15 @@ class Dropzone extends React.Component {
...
     return accepts(file, this.props.accept);
   }
 
-  fileMatchSize(file) {
+  fileMatchSize(file: Object) {
     return file.size <= this.props.maxSize && file.size >= this.props.minSize;
   }
 
-  allFilesAccepted(files) {
+  allFilesAccepted(files: Array<Object>) {
```
Comment (okonet) [inline src/index.js]: Let's define the `DataTransferItemType` based on https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem and reuse it. Also this method return type is missing (`:boolean`)

File: .babelrc (hunk 5 above)
Comment (ryyppy) [inline .babelrc]: hm,... depends on your dependencies... if there are no dependencies, it's easier to consume... otherwise we should publish separate libdefs on flow-typed for now

File: src/index.js (hunk 9 above)
Comment (ming-soon) [inline src/index.js]: @okonet I think we should use the `File` type based on [https://developer.mozilla.org/en-US/docs/Web/API/File](https://developer.mozilla.org/en-US/docs/Web/API/File)

File: src/index.js (hunk 7 above)
Comment (ming-soon) [inline src/index.js]: This can be `HTMLInputElement`.

File: .babelrc (hunk 5 above)
Comment (ming-soon) [inline .babelrc]: @okonet By default, the preset [react](https://babeljs.io/docs/plugins/preset-react/) includes the `transform-flow-strip-types` plugin. And right now it is not possible to disable a specific plugin from preset, AFAIK.
[http://stackoverflow.com/questions/33707189/is-there-an-option-to-disable-a-plugin-in-babel-when-using-a-preset](http://stackoverflow.com/questions/33707189/is-there-an-option-to-disable-a-plugin-in-babel-when-using-a-preset)

File: src/index.js [hunk 10]
```diff
@@ -372,4 +384,31 @@ Dropzone.defaultProps = {
   minSize: 0
 };
 
+Dropzone.propTypes = {
```
Comment (okonet) [inline src/index.js]: Hmm, the whole point of having Flow types is to remove PropTypes. I'm wondering why did you add them back?

Comment (okonet) [inline src/index.js]: Hmm, can't see this change for the whole PR but only for the last commit. So, I guess, nevermind my comment.

**Comment by SaraVieira** [comment]
@okonet why hasn't this been merged?

**Comment by okonet** [comment]
I wanted to remove style management before proceeding with this one. But it's taking too much time so I'd appreciate any help

**Comment by SaraVieira** [comment]
In this or style management?

**Comment by okonet** [comment]
@SaraVieira In both I guess since after removing style management this one will need to be updated.

File: src/index.js [hunk 11]
```diff
@@ -316,22 +393,19 @@ Dropzone.propTypes = {
...
   rejectClassName: React.PropTypes.string, // className for rejected state
-
   disablePreview: React.PropTypes.bool, // Enable/disable preview generation
   disableClick: React.PropTypes.bool, // Disallow clicking on the dropzone container to open file dialog
   onFileDialogCancel: React.PropTypes.func, // Provide a callback on clicking the cancel button of the file dialog
-
   inputProps: React.PropTypes.object, // Pass additional attributes to the <input type="file"/> tag
   multiple: React.PropTypes.bool, // Allow dropping multiple files
-  accept: React.PropTypes.string, // Allow specific types of files. See https://github.com/okonet/attr-accept for more information
+  accept: React.PropTypes.string, // Allow specific types of files. See https://github.com/okonet/attrccept for more information
```
Comment (b0gok) [inline src/index.js]: @okonet `https://github.com/okonet/attrccept` is invalid link and `https://github.com/okonet/attr-accept` is valid. Is it correct change?

Comment (okonet) [reply src/index.js]: No, this is incorrect. Thanks for mentioning.

**Comment by rolandjitsu** [comment]
Is anyone still gonna work on this? And if so, how will this play with TypeScript? Because I plan on migrating the current codebase to TS after https://github.com/react-dropzone/react-dropzone/pull/717.

**Comment by capaj** [comment]
@rolandjitsu you can't have both flow and typescript in the same project. 
I'd say close this PR and move on with typescript. TS has become the standard for typing in JS community. Even projects like yarn or jest which were the biggest pushers of flow migrated to TS.

**Comment by rolandjitsu** [comment]
@okonet what do you think, should we close this PR?

**Comment by rxmarbles** [comment]
@rolandjitsu I think it's not worth the effort and agree it should be closed. However if we do want to some form of typing we could look to switch to typscript

**Comment by rolandjitsu** [comment]
@rxmarbles ok. We already have TS typings, but I also have started working on migrating the current codebase to TS.

**Comment by rxmarbles** [comment]
Ah yes totally forgot about that 😂. Sweet. Look forward to reviewing that PR.
